### PR TITLE
MEPTS-1012 | TPT INITIATION CHANGES

### PR DESCRIPTION
**Pull request by TinoMuianga**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 1]
```diff
@@ -624,4 +632,789 @@ public CohortDefinition getPatientsIPTStart5WithTPTAndMarkedFILT() {
...
+            + "  p.voided = 0 "
+            + "  AND e.voided = 0 "
+            + "  AND o.voided = 0 "
+            + "  AND o2.voided = 0 "
+            + "  AND e.location_id = :location "
+            + "  AND o.concept_id = ${23985} "
+            + "  AND o.value_coded IN (${23954}, ${23984}) "
+            + "  AND o2.concept_id = ${23987} "
+            + "  AND o2.value_coded IN (${1257}, ${1267}) "
+            + "  OR o2.value_coded IS NULL "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: Check this condition. It doesn't seem to implement the specification correctly

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 2]
```diff
@@ -624,4 +632,789 @@ public CohortDefinition getPatientsIPTStart5WithTPTAndMarkedFILT() {
...
+            + "                  INNER JOIN obs o ON e.encounter_id = o.encounter_id "
+            + "                  INNER JOIN obs o2 ON e.encounter_id = o2.encounter_id "
+            + "              WHERE p.voided = 0 AND e.voided = 0 "
+            + "                  AND o.voided = 0 "
+            + "                  AND e.encounter_type=  ${60} "
+            + "                  AND (o.concept_id=  ${23985}   AND o.value_coded IN ( ${23954}  , ${23984} )) "
+            + "                  AND (o2.concept_id=  ${23987}   AND o2.value_coded IN ( ${1256} , ${1705} )) "
+            + "                  AND e.encounter_datetime BETWEEN :startDate AND :endDate "
+            + "       "
+            + "       "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: Remove whitespaces

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 3]
```diff
@@ -624,4 +632,789 @@ public CohortDefinition getPatientsIPTStart5WithTPTAndMarkedFILT() {
...
+            + "  AND e.encounter_datetime BETWEEN :startDate "
+            + "  AND :endDate "
+            + "  AND p.patient_id NOT IN ( "
+            + "    SELECT "
+            + "      p.patient_id "
+            + "    FROM "
+            + "      patient p "
+            + "      INNER JOIN encounter e ON e.patient_id = p.patient_id "
+            + "      INNER JOIN obs o ON e.encounter_id = o.encounter_id "
+            + "      INNER JOIN (  SELECT p.patient_id, e.encounter_datetime AS start_date "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: The system will consider the oldest date amongst all sources as the 3HP Start Date - This note wasn't implemented

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 4]
```diff
@@ -624,4 +632,789 @@ public CohortDefinition getPatientsIPTStart5WithTPTAndMarkedFILT() {
...
+            + "WHERE  p.voided = 0 "
+            + "       AND e.voided = 0 "
+            + "       AND o.voided = 0 "
+            + "       AND o2.voided = 0 "
+            + "       AND o3.voided = 0 "
+            + "       AND e.location_id = :location "
+            + "       AND e.encounter_type = ${6} "
+            + "       AND o.concept_id = ${6122} "
+            + "       AND o.value_coded = ${1256} "
+            + "        OR o2.concept_id = ${23985} "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: The patient must have had both 'Profilaxia TPT' and 'Estado Profilaxia' to be eligible

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 5]
```diff
@@ -624,4 +632,789 @@ public CohortDefinition getPatientsIPTStart5WithTPTAndMarkedFILT() {
...
+    cd.addParameter(new Parameter("endDate", "endDate", Date.class));
+    cd.addParameter(new Parameter("location", "location", Location.class));
+
+    Map<String, Integer> valuesMap = new HashMap<>();
+    valuesMap.put("6", hivMetadata.getAdultoSeguimentoEncounterType().getEncounterTypeId());
+    valuesMap.put("9", hivMetadata.getPediatriaSeguimentoEncounterType().getEncounterTypeId());
+    valuesMap.put("6128", hivMetadata.getDataInicioProfilaxiaIsoniazidaConcept().getConceptId());
+
+    String query =
+        "SELECT p.patient_id "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: This query exists in the code, you can reuse it

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 6]
```diff
@@ -624,4 +632,789 @@ public CohortDefinition getPatientsIPTStart5WithTPTAndMarkedFILT() {
...
+            + "                          AND e.voided = 0 "
+            + "                          AND o.voided = 0 "
+            + "                          AND o2.voided = 0 "
+            + "                          AND e.location_id = :location "
+            + "                          AND e.encounter_type = ${60} "
+            + "                          AND o.concept_id = ${23985} "
+            + "                          AND o.value_coded IN ( ${656}, ${23982} ) "
+            + "                          AND o2.concept_id = ${23987} "
+            + "                          AND o2.value_coded = ${1257} "
+            + "                           OR o2.value_coded IS NOT NULL "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: This preposition isn't according to specification

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 7]
```diff
@@ -624,4 +632,789 @@ public CohortDefinition getPatientsIPTStart5WithTPTAndMarkedFILT() {
...
+            + "                                FROM   patient p "
+            + "                                       inner join encounter e "
+            + "                                               ON p.patient_id = e.patient_id "
+            + "                                       inner join obs o "
+            + "                                               ON e.encounter_id = "
+            + "                                                  o.encounter_id "
+            + "                                WHERE  p.voided = 0 "
+            + "                                       AND e.voided = 0 "
+            + "                                       AND o.voided = 0 "
+            + "                                       AND e.location_id = :location "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: Specify the encounter type in the exclusions queries

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 8]
```diff
@@ -624,4 +632,789 @@ public CohortDefinition getPatientsIPTStart5WithTPTAndMarkedFILT() {
...
+    valuesMap.put("23985", tbMetadata.getRegimeTPTConcept().getConceptId());
+    valuesMap.put("656", tbMetadata.getIsoniazidConcept().getConceptId());
+    valuesMap.put("23982", tbMetadata.getIsoniazidePiridoxinaConcept().getConceptId());
+    valuesMap.put("23987", hivMetadata.getPatientTreatmentFollowUp().getConceptId());
+    valuesMap.put("1256", hivMetadata.getStartDrugs().getConceptId());
+    valuesMap.put("1705", hivMetadata.getRestartConcept().getConceptId());
+    valuesMap.put("60", tbMetadata.getRegimeTPTEncounterType().getEncounterTypeId());
+
+    String query =
+        "SELECT p.patient_id "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: The system will consider the oldest date amongst all sources as the IPT Start Date

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 9]
```diff
@@ -143,62 +232,103 @@ public DataDefinition getPatientsReceivedTPTInTheLastFollowUpConsultation() {
...
+            + "       AND e.encounter_type = ${6} "
+            + "       AND e.voided = 0 "
+            + "       AND o.voided = 0 "
+            + "       AND ( o.concept_id = ${6122} "
+            + "             AND o.value_coded IN ( ${1256}, ${1257} ) ) "
+            + "        OR ( ( o2.concept_id = ${23985} "
+            + "               AND o2.value_coded IN ( ${656}, ${23954} ) ) "
+            + "             AND ( o3.concept_id = ${165308} "
+            + "                   AND o3.value_coded IN ( ${1256}, ${1257} ) ) ) "
+            + "           AND e.encounter_datetime < Curdate() "
```
Comment (eliazmuchave) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]: remove this condition

**Review by eliazmuchave — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java [hunk 10]
```diff
@@ -788,77 +790,76 @@ public CohortDefinition getPatientsWithFirstFiltRegimeTpt() {
...
+            + "                    FROM patient pp "
+            + "                            INNER JOIN encounter ee ON ee.patient_id = pp.patient_id "
+            + "                            INNER JOIN obs oo ON oo.encounter_id = ee.encounter_id "
+            + "                    WHERE pp.voided = 0 "
+            + "                    AND p.patient_id = pp.patient_id "
+            + "                    AND ee.voided = 0 "
+            + "                    AND oo.voided = 0 "
+            + "                    AND ee.location_id = :location "
+            + "                    AND ee.encounter_type = 60 "
+            + "                    AND oo.concept_id = 23985 "
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: Please map the concepts properly. ${60}

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 11]
```diff
@@ -416,19 +418,11 @@ public DataDefinition getPatientAnd3HPInitiationDateOnFichaClinica() {
...
-            + "WHERE  p.voided = 0 "
-            + "       AND e.voided = 0 "
-            + "       AND o.voided = 0 "
-            + "       AND o2.voided = 0 "
-            + "       AND o3.voided = 0 "
+            + "       INNER JOIN encounter e ON p.patient_id = e.patient_id "
+            + "       INNER JOIN obs o ON e.encounter_id = o.encounter_id "
+            + "       INNER JOIN obs o2 ON e.encounter_id = o2.encounter_id "
+            + "       INNER JOIN obs o3 ON e.encounter_id = o3.encounter_id "
+            + "WHERE  p.voided = 0 AND e.voided = 0 AND o.voided = 0 AND o2.voided = 0 AND o3.voided = 0 "
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java:415]: check this implementation

this two conditions should happen on the same encounter 

``(o2.concept_id = ${23985}) AND (o2.value_coded = ${23954})) ) "
            + "           AND o3.concept_id = ${165308} AND o3.value_coded = ${1256} ``

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 12]
```diff
@@ -632,24 +638,51 @@ public DataDefinition getPatientsAndLast3HPTypeOfDispensation() {
...
             + "                         AND e.location_id = :location "
             + "                         AND e.encounter_type = ${6} "
-            + "                         AND o.concept_id = ${1719} AND o.value_coded IN ( ${23954} ) "
-            + "                         AND e.encounter_datetime >= :startDate "
-            + "                         AND e.encounter_datetime <= :endDate "
-            + "                  GROUP  BY p.patient_id) 3hp "
-            + "               ON 3hp.patient_id = p.patient_id "
+            + "                         AND ( ( o.concept_id = ${1719} AND o.value_coded = ${23954} ) "
+            + "                                OR ( o2.concept_id = ${23985} AND o2.value_coded = ${23954} ) ) "
+            + "                         AND o3.concept_id = ${165308} AND o3.value_coded = ${1256} "
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]: apply the same changes on this implementation

`AND ( ( o.concept_id = ${1719} AND o.value_coded = ${23954} ) "
            + "                                OR ( o2.concept_id = ${23985} AND o2.value_coded = ${23954} ) ) "
            + "                         AND o3.concept_id = ${165308} AND o3.value_coded = ${1256} "`

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java [hunk 13]
```diff
@@ -689,36 +724,68 @@ public DataDefinition get3HPCompletionDateonFichaResumo() {
...
+            + "                               INNER JOIN encounter e ON p.patient_id = e.patient_id "
+            + "                               INNER JOIN obs o ON e.encounter_id = o.encounter_id "
+            + "                               INNER JOIN obs o2 ON e.encounter_id = o2.encounter_id "
+            + "                               INNER JOIN obs o3 ON e.encounter_id = o3.encounter_id "
+            + "                        WHERE  p.voided = 0 AND e.voided = 0 AND o.voided = 0 AND o2.voided = 0 AND o3.voided = 0 "
+            + "                               AND e.location_id = :location "
+            + "                               AND e.encounter_type = ${6} "
+            + "                               AND ( ((o.concept_id = ${1719}) AND (o.value_coded = ${23954})) "
+            + "                               OR ((o2.concept_id = ${23985}) AND (o2.value_coded = ${23954})) ) "
+            + "                                   AND o3.concept_id = ${165308} AND o3.value_coded = ${1256} "
```
Comment (helderjosue) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationDataDefinitionQueries.java]: please check this 
`AND ( ((o.concept_id = ${1719}) AND (o.value_coded = ${23954})) "
            + "                               OR ((o2.concept_id = ${23985}) AND (o2.value_coded = ${23954})) ) "
            + "                                   AND o3.concept_id = ${165308} AND o3.value_coded = ${1256} "`

**Review by helderjosue — COMMENTED** [review]
please check the comments

**Review by helderjosue — APPROVED** [review]
It looks good to me!

**Comment by ningosi** [comment]
@eliazmuchave the changes you requested here, were they resolved? if yes, you might need to approve them so that I have a final look at the PR

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java (hunk 1 above)
Comment (eliazmuchave) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TPTInitiationCohortQueries.java]: Implemented

**Review by ningosi — APPROVED** [review]

**Review by eliazmuchave — APPROVED** [review]

**Pull request merged**